Crypto/CertificateTest.cc

```diff
@@ -194,12 +197,168 @@ TEST_CASE("Persistent key and cert", "[Certs]") {
     CHECK(csr2->subjectName() == kSubjectName);
     CHECK(csr2->subjectPublicKey()->data(KeyFormat::Raw) == key->publicKey()->data(KeyFormat::Raw));
 
-    // Make the cert persistent:
-    cert->makePersistent();
-    Retained<Cert> cert2 = Cert::load(pubKey);
-    REQUIRE(cert2);
-    CHECK(cert2->data() == cert->data());
+    // Delete the cert to cleanup:
+    Cert::deleteCert("cert1");
+    CHECK(Cert::loadCert("cert1") == nullptr);
+    
+    // Save the cert:
+    cert->save("cert1", true);
+    
+    // Load the cert with the persistent ID:
+    Retained<Cert> certA = Cert::loadCert("cert1");
+    REQUIRE(certA);
+    CHECK(certA->data() == cert->data());
+    
+    // Load the cert with public key:
+    Retained<Cert> certB= Cert::load(pubKey);
+    REQUIRE(certB);
+    CHECK(certB->data() == cert->data());
+    
+    // Delete the cert:
+    Cert::deleteCert("cert1");
+    CHECK(Cert::loadCert("cert1") == nullptr);
 }
+
+
+TEST_CASE("Persistent save duplicate cert or id", "[Certs]") {
+    // Create a keypair and a cert1:
+    Retained<PersistentPrivateKey> key1 = PersistentPrivateKey::generateRSA(2048);
+    Retained<PublicKey> pubKey = key1->publicKey();
+    CHECK(pubKey != nullptr);
+
+    Cert::IssuerParameters issuerParams1;
+    issuerParams1.serial = "1"_sl;
+    issuerParams1.validity_secs = 3600*24;
+    Retained<Cert> cert1 = new Cert(DistinguishedName(kSubjectName), issuerParams1, key1);
+
+    // Delete cert1 to cleanup:
+    Cert::deleteCert("cert1");
+    
+    // Save cert1:
+    cert1->save("cert1", true);
+    Retained<Cert> cert1a = Cert::loadCert("cert1");
+    REQUIRE(cert1a);
+    CHECK(cert1a->data() == cert1->data());
+    
+    // Save cert1 again with the same id:
+    ExpectException(error::LiteCore, error::CryptoError, [&]{
+        cert1->save("cert1", true);
+    });
+        
+#ifdef __APPLE__
+    // Save cert1 again with a different id:
+    ExpectException(error::LiteCore, error::CryptoError, [&]{
+        cert1->save("cert2", true);
+    });
+#endif
+    
+    // Create another keypair and cert2:
+    Retained<PersistentPrivateKey> key2 = PersistentPrivateKey::generateRSA(2048);
+    Retained<PublicKey> pubKey2 = key2->publicKey();
+    CHECK(pubKey2 != nullptr);
+
+    Cert::IssuerParameters issuerParams2;
+    issuerParams2.serial = "2"_sl;
+    issuerParams2.validity_secs = 3600*24;
+    Retained<Cert> cert2 = new Cert(DistinguishedName(kSubject2Name), issuerParams2, key1);
+    
+    // Delete cert2 to cleanup:
+    Cert::deleteCert("cert2");
+    
+    // Save cert2 to an existing ID:
+    ExpectException(error::LiteCore, error::CryptoError, [&]{
+        cert2->save("cert1", true);
+    });
+    
+    // Save cert2:
+    cert2->save("cert2", true);
+    Retained<Cert> cert2a = Cert::loadCert("cert2");
+    REQUIRE(cert2a);
+    CHECK(cert2a->data() == cert2->data());
+    
+    // Delete cert1:
+    Cert::deleteCert("cert1");
+    CHECK(Cert::loadCert("cert1") == nullptr);
```

Comment:
is it worth checking, save with the deleted certificate identifier, once its deleted?

Reply:
I think so. I will add the test when the core implementation is reviewed.

Reply:
Done. I added the case into the other test which should cover this case. Thanks.